nucypher/blockchain/eth/utils.py

```diff
@@ -14,33 +14,247 @@
 You should have received a copy of the GNU Affero General Public License
 along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
 """
+
+
+from decimal import Decimal
+
+import eth_utils
 import maya
+from nacl.hash import sha256
+from typing import Union, Tuple
 
+from nucypher.blockchain.eth.agents import NucypherTokenAgent
 from nucypher.blockchain.eth.constants import (MIN_ALLOWED_LOCKED,
                                                MAX_ALLOWED_LOCKED,
                                                MIN_LOCKED_PERIODS,
                                                MAX_MINTING_PERIODS,
                                                SECONDS_PER_PERIOD)
 
 
+def datetime_to_period(datetime: maya.MayaDT) -> int:
+    """Converts a MayaDT instance to a period number."""
+    future_period = datetime._epoch // int(SECONDS_PER_PERIOD)
+    return int(future_period)
+
+
+def datetime_at_period(period: int) -> maya.MayaDT:
+    """Returns the datetime object at a given period, future, or past."""
+
+    now = maya.now()
+    current_period = datetime_to_period(datetime=now)
+    delta_periods = period - current_period
+
+    # +
+    if delta_periods:
+        target_period = now + maya.timedelta(days=delta_periods)
+
+    # -
+    else:
+        target_period = now - maya.timedelta(days=delta_periods)
+
+    return target_period
+
+
+def calculate_period_duration(future_time: maya.MayaDT) -> int:
+    """Takes a future MayaDT instance and calculates the duration from now, returning in periods"""
+    future_period = datetime_to_period(datetime=future_time)
+    current_period = datetime_to_period(datetime=maya.now())
+    periods = future_period - current_period
+    return periods
+
+
+class NU:
+    """
+    An amount of NuCypher tokens that doesn't hurt your eyes.
+
+    The easiest way to use NU, is to pass an int, and denomination string:
+
+    nu = NU(100, 'NU')
+    nu_wei = NU(15000000000000000000000, 'NUWei')
+
+    alternately
+
+    nu = NU.from_tokens(100)
+    nu_wei = NU.from_nu_wei(15000000000000000000000)
+
+    Token quantity is stored internally as a Decimal in the smallest denomination,
+    and all arithmetic operations use this value.
+    """
+
+    __symbol = 'NU'
+    __decimals = 18
+    __agent_class = NucypherTokenAgent
+
+    # conversions to smallest denomination
+    __denominations = {'NUWei': 10 ** __decimals,
+                       'NU': 1}
+
+    def __init__(self, value: int, denomination: str):
+
+        # Calculate smallest denomination and store it
+        divisor = self.__denominations[denomination]
+        self.__value = Decimal(value) / divisor
```

Comment:
I think the use of `Decimal` is only justified if, well, we're going to use decimals. However, the type hint in the constructor suggests that the input value is an `int`. Perhaps that should be changed to `Union[int, str, float]`.

Another thing is that the use of `Decimal` can be problematic due to special values ('NaN', 'Infinity', etc.). Mix that with the lack of type checking and we can have awful things like:
```python
>>> Decimal('Infinity') * 10
Decimal('Infinity')
>>> Decimal('NaN') / 100
Decimal('NaN')
>>> Decimal('NaN') == Decimal('NaN')
False
```
I think the NU class requires validating the input with [`Decimal.is_normal()`](https://docs.python.org/3/library/decimal.html#decimal.Decimal.is_normal). There are also some problems with floats and strings that represent floats. I've added some test for edge cases on top of your PR.

Reply:
Excellent suggestions, I honestly do not know why it escaped me to consider floats :joy:

Reply:
(Moving to PR #830 )

Reply:
:heavy_check_mark: